File: SMB_ENUM/python/smb_enum.py
#!/usr/bin/env python3
import subprocess
import sys
import re

def check_root():
    if os.geteuid() != 0:
        print("[-] Requires root privileges for hash extraction")
        sys.exit(1)

def get_workgroups():
    print("[+] Discovering workgroups...")
    result = subprocess.run(["nmblookup", "-S", "__SAMBA__"], capture_output=True, text=True)
    workgroups = re.findall(r'<GROUP>\s+(\S+)', result.stdout)
    return sorted(set(workgroups))

def enum_users(workgroups):
    print("\n[+] Enumerating users...")
    users = set()
    for wg in workgroups:
        cmd = f"rpcclient -U% -W '{wg}' -c 'enumdomusers' 127.0.0.1"
        result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
        matches = re.findall(r'\[(.*?)\]', result.stdout)
        users.update(matches)
    
    print("Found Users:")
    for user in sorted(users):
        print(f" - {user}")

def extract_hashes():
    print("\n[+] Extracting SMB hashes...")
    try:
        hashes = subprocess.check_output(
            ["pdbedit", "-L", "-w"], 
            stderr=subprocess.DEVNULL, 
            text=True
        )
        for line in hashes.splitlines():
            if ":" in line and not line.endswith(":$"):
                print(line)
    except Exception as e:
        print(f"[-] Hash extraction failed: {str(e)}")

if __name__ == "__main__":
    check_root()
    wgs = get_workgroups()
    print("Workgroups Found:\n" + "\n".join(wgs))
    enum_users(wgs)
    extract_hashes()

File: SMB_ENUM/C/smb_enum.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAXBUF 4096

void check_root() {
    if (geteuid() != 0) {
        fprintf(stderr, "[-] Must be run as root\n");
        exit(EXIT_FAILURE);
    }
}

void get_workgroups() {
    printf("[+] Discovering workgroups...\n");
    FILE *fp = popen("nmblookup -S __SAMBA__ | grep -oP '<GROUP>\\s+\\K\\S+' | sort -u", "r");
    char buffer[MAXBUF];
    
    printf("Workgroups:\n");
    while (fgets(buffer, MAXBUF, fp)) {
        printf(" - %s", buffer);
    }
    pclose(fp);
}

void enum_users() {
    printf("\n[+] Enumerating users...\n");
    system("for wg in $(nmblookup -S __SAMBA__ | grep -oP '<GROUP>\\s+\\K\\S+' | sort -u); do "
           "echo \"=== Workgroup: $wg ===\"; "
           "rpcclient -U% -W \"$wg\" -c \"enumdomusers\" 127.0.0.1 2>/dev/null | "
           "grep -oP '\\[.*?\\]' | tr -d '[]'; done | sort -u");
}

void extract_hashes() {
    printf("\n[+] SMB Hashes:\n");
    system("pdbedit -L -w | grep -v ':\\$' | column -t -s:");
}

int main() {
    check_root();
    get_workgroups();
    enum_users();
    extract_hashes();
    return 0;
}
